flink-connector-bigquery-common/src/main/java/com/google/cloud/flink/bigquery/common/utils/AvroToBigQuerySchemaTransform.java

```diff
@@ -0,0 +1,456 @@
+     * sub-fields, and then constructs a BigQuery record field with the converted fields.
+     *
+     * @param avroSchema the Avro schema of the record field
+     * @param name the name of the field
+     * @param nestedLevel the current nesting level of the field
+     * @return the converted BigQuery record field with nested fields.
+     */
+    private static Field convertAvroRecordFieldToBigQueryField(
+            org.apache.avro.Schema avroSchema, String name, int nestedLevel) {
+        List<Field> fields = new ArrayList<>();
+        Preconditions.checkState(
+                !avroSchema.getFields().isEmpty(),
+                "The avro record schema must have at least one field.");
+        // Iterate over each record field and obtain the nested record fields.
+        for (int i = 0; i < avroSchema.getFields().size(); i++) {
+            org.apache.avro.Schema nestedAvroSchema = avroSchema.getFields().get(i).schema();
+            fields.add(
+                    convertAvroFieldToBigQueryField(
+                            nestedAvroSchema,
+                            avroSchema.getFields().get(i).name(),
+                            nestedLevel + 1));
+        }
+        FieldList nestedBigQueryFields = FieldList.of(fields);
+        return Field.newBuilder(name, LegacySQLTypeName.RECORD, nestedBigQueryFields)
+                .setDescription((avroSchema.getDoc() != null) ? avroSchema.getDoc() : null)
+                .setMode(Field.Mode.REQUIRED)
+                .build();
+    }
+
+    /**
+     * Helper function to convert the UNION type schema field to a BigQuery Field.
+     *
+     * <p>BigQuery supports nullable fields but not unions with multiple non-null types. This method
+     * validates the Avro UNION schema to ensure it conforms to BigQuery's requirements.
+     *
+     * <p>Valid UNION schemas are:
+     *
+     * <ul>
+     *   <li>["null", datatype] - Represents a nullable field.
+     *   <li>["datatype", null] - Represents a nullable field.
+     *   <li>[datatype] - Represents a non-nullable field.
+     * </ul>
+     *
+     * <p>Invalid UNION schemas include:
+     *
+     * <ul>
+     *   <li>["null"] - Only null type is not supported.
+     *   <li>[datatype1, datatype2, ...] - Unions without a null type are not supported.
+     *   <li>["null", ARRAY] Arrays in BigQuery cannot be nullable
+     *   <li>[ARRAY, "null"] Arrays in BigQuery cannot be nullable
+     * </ul>
+     *
+     * <p>If the UNION schema is valid, this method returns a BigQuery field with the schema of the
+     * non-null datatype. Otherwise, it throws an {@link UnsupportedOperationException}.
+     *
+     * @param avroSchema The Avro UNION schema to process.
+     * @return The BigQuery field corresponding to the non-null type in the UNION.
+     * @throws UnsupportedOperationException if the UNION schema is invalid for BigQuery.
+     */
+    private static Field convertAvroNullableFieldToBigQueryField(
```

Comment:
Refactor this method as perviously hinted

Reply:
Done